geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/Z3IdxStrategy.scala

```diff
@@ -205,18 +206,29 @@ object Z3IdxStrategy extends StrategyProvider {
   val Z3_ITER_PRIORITY = 21
   val FILTERING_ITER_PRIORITY = 25
 
+  override protected def statsBasedCost(sft: SimpleFeatureType,
+                                        filter: QueryFilter,
+                                        transform: Option[SimpleFeatureType],
+                                        stats: GeoMesaStats): Option[Long] = {
+    // https://geomesa.atlassian.net/browse/GEOMESA-1166
+    // TODO check date range and use z2 instead if too big
+    // TODO also if very small bbox, z2 has ~10 more bits of lat/lon info
+    filter.singlePrimary match {
+      case Some(f) => stats.getCount(sft, f, exact = false)
+      case None    => Some(Long.MaxValue)
+    }
+  }
+
   /**
-   * Gets the estimated cost of running the query. Currently, cost is hard-coded to sort between
-   * strategies the way we want. Z3 should be more than id lookups (at 1), high-cardinality attributes (at 1)
-   * and less than STidx (at 400) and unknown cardinality attributes (at 999).
-   *
-   * Eventually cost will be computed based on dynamic metadata and the query.
-   */
-  override def getCost(filter: QueryFilter, sft: SimpleFeatureType, hints: StrategyHints) =
-  // https://geomesa.atlassian.net/browse/GEOMESA-1166
-  // TODO check date range and use z2 instead if too big
-  // TODO also if very small bbox, z2 has ~10 more bits of lat/lon info
-    if (filter.primary.exists(isSpatialFilter)) 200 else 401
+    * More than id lookups (at 1), high-cardinality attributes (at 1).
+    * Less than unknown cardinality attributes (at 999).
+    * With a spatial component, less than z2, otherwise more than z2 (at 400)
+    */
+  override protected def indexBasedCost(sft: SimpleFeatureType,
+                                        filter: QueryFilter,
+                                        transform: Option[SimpleFeatureType]): Long = {
+    if (filter.primary.exists(isSpatialFilter)) 200L else 401L
```

Comment:
yeah, these are the old way we calculated costs - left for data stores that don't have stats, or if the stats query planning proves disastrous, you can fall back to this with a system property.